searches/binary_search.nim

```diff
@@ -0,0 +1,99 @@
+## Binary Search
+## =============
+## Binary search is an efficient searching algorithm. It is faster than linear 
+## search except for small arrays. However, binary search can only be used on sorted arrays.
+## If the array is not sorted, it must be sorted first before binary search can be applied.
+##
+## Binary search starts by comparing the key value with the middle element of the array.
+## If the key value matches the middle element, the search is complete.
+## If the key value is less than the middle element, the search continues on the lower half of the array.
+## If the key value is greater than the middle element, the search continues on the upper half of the array.
+## This process repeats until the middle element matches the key value or the search space is exhausted.
+## https://en.wikipedia.org/wiki/Binary_search_algorithm
+##
+## Time Complexity: O(log n), where n is the length of the array.
+## Space Complexity in iterative approach: O(1)
+## Space Complexity in recursive approach: O(n)
+{.push raises: [].}
+
+runnableExamples:
+  import std/options
+
+  var arr1 = [0, 1, 2, 4, 5, 6]
+  doAssert binarySearchIterative(arr1, 5) == some(Natural(4))
+
+  var arr2 = ['a', 'c', 'd', 'e', 'x', 'z']
+  doAssert binarySearchIterative(arr2, 'e') == some(Natural(3))
+
+  var arr3 = [0, 1, 2, 3, 4]
+  doAssert binarySearchIterative(arr3, 7) == none(Natural)
+
+import std/options
+
+func binarySearchIterative*[T:Ordinal](arr: openArray[T], key: T): Option[Natural] =
+  ## Binary search can only be applied to sorted arrays.
+  ## For this function array should be sorted in ascending order.
+  var 
+    left = arr.low
+    right = arr.high
+
+  while left <= right:
+    let mid = left + (right - left) div 2
+
+    if arr[mid] == key: 
+      return some(Natural(mid))
+
+    if key < arr[mid]: 
+      right = mid - 1
+    else:
+      left = mid + 1
+  # `none(Natural)` is returned when both halfs are empty after some iterations.
+  none(Natural)
+
+func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
```

Comment:
```suggestion
func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: Natural, key: T): Option[Natural] =
```

Reply:
Natural here really only adds the complexity as `right` could be -1 (when `mid` = 0). Negative indices can be easily handled by the algorithm without complex edge cases.

Reply:
> Natural here really only adds the complexity as `right` could be -1 (when `mid` = 0). I believe, negative indices can be easily handled by the algorithm without complex edge cases.

There's no negative indexes for `openArrays` and they are necessary to be array/seq agnostic. The other options is to use a concept that will embrace ordinal-indexed types or write type-specific overloads.